jetty-util/src/main/java/org/eclipse/jetty/util/PathUtil.java

```diff
@@ -0,0 +1,157 @@
+//
+// ========================================================================
+// Copyright (c) 1995-2021 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.util;
+
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+
+/**
+ * Collection of Path encoding / decoding utilities for URI, Http Headers, and HTML
+ */
+public class PathUtil
+{
+    // Simple hex array
+    private static final char[] HEXES = "0123456789ABCDEF".toCharArray();
+    // The first 127 characters and their encodings.
+    private static final String[] URI_ENCODED_CHARS;
+
+    static
+    {
+        URI_ENCODED_CHARS = new String[127];
+        Arrays.fill(URI_ENCODED_CHARS, null);
+
+        // URI Reserved Chars
+        String uriReservedGenDelims = ":?#[]@"; // intentionally missing "/"
+        String uriReservedSubDelims = "!$&'()*+,;=";
+        // Extra Reserved Chars (specified by Jetty)
+        String jettyReserved = "%\"<> \\^`{}|";
+
+        String reserved = uriReservedGenDelims + uriReservedSubDelims + jettyReserved;
+
+        for (int i = 0; i < URI_ENCODED_CHARS.length; i++)
+        {
+            if ((i < 0x20) || // control characters
+                (reserved.indexOf(i) != (-1)))
+            {
+                // encoding needed
+                URI_ENCODED_CHARS[i] = String.format("%%%02X", i);
+            }
+        }
+    }
+
+    public static String encodePath(String rawpath)
+    {
+        byte[] rawpathbuf = rawpath.getBytes(StandardCharsets.UTF_8);
+        int len = rawpathbuf.length;
+        StringBuilder buf = new StringBuilder(len * 2);
+        for (byte c : rawpathbuf)
+        {
+            if ((c >= 0) && c <= URI_ENCODED_CHARS.length)
+            {
+                // this is the 7-bit US-ASCII space
+                // only some of these characters are encoded
+                String e = URI_ENCODED_CHARS[c];
+                if (e != null)
+                    buf.append(e);
+                else
+                    buf.append((char)c);
+            }
+            else
+            {
+                // all of these characters are 8-bit and above (unicode) and should be encoded.
+                buf.append('%');
+                buf.append(HEXES[(c & 0xF0) >> 4]);
+                buf.append(HEXES[(c & 0x0F)]);
+            }
+        }
+        return buf.toString();
+    }
+
+    public static String encodePathDelayAlloc(String rawpath)
```

Comment:
I think the JMH has indicated enough that this is a better algorithm than the `URIUtil.encodePath`.   Just update that method and don't need to check in and maintain the jmh or alternatives.

If we do want to check in a JMH, it should be a general comparison between doing a static array lookup vs `switch(char)`

Reply:
It would be interesting to jmh compare the approach used in HttpParser, which looks up characters in an array of `Token[256]` and then does a switch on the enum obtained from `token.getType`.